interface/src/app/signature/enriched_signatures.js

```diff
@@ -0,0 +1,195 @@
+        var message = errGen('Failed to get signatures: ', err);
+        $log.error(message);
+        self.statusMessage = message + '. Please try again later.';
+      }
+    );
+
+    // For the moment, the total number of genes in our universe
+    // to perform this enrichment analysis, is the total number of
+    // Entrez IDs in our database for Pseudomonas aeruginosa.
+    Gene.get(
+      {},
+      function success(response) {
+        self.geneNum = response.meta.total_count;
+      },
+      function error(err) {
+        var message = errGen('Failed to get total gene number: ', err);
+        $log.error(message);
+        self.statusMessage = message + '. Please try again later.';
+      }
+    );
+
+    self.pValueCutoff = 0.05;
+    var pValueSigDigits = 3;
+
+    // This is an object, of signatures per participation type
+    var signtrsPerPT = {};
+
+    self.enrichedSignatures = {};
+
+    // This is the main function that calculates the signature enrichments.
+    // It calculates the enrichment for each signature that has participatory
+    // genes which were in the url genes (the genes searched).
+    var calculateEnrichments = function(totalGeneNum, signature, cutoff) {
+      // For the moment, this is the total number of Entrez IDs for
+      // Pseudomonas in the database.
+      var N = totalGeneNum;
+
+      // This will be the number of genes in the url (which are the genes the
+      // user selected in the previous page)
+      var m = userSelectedGenes.length;
+
+      var pValueArray = [];
+
+      var signatureIds = Object.keys(signatures);
+
+      for (var i = 0; i < signatureIds.length; i++) {
+        var signatureId = signatureIds[i];
+        var genes = signatures[signatureId];
+        var n = genes.length;
+        var k = 0;
+
+        for (i = 0; i < userSelectedGenes.length; i++) {
+          var selectedGene = userSelectedGenes[i];
+          if (genes.indexOf(selectedGene) > -1) {
+            k++;
+          }
+        }
+
+        var pValue = 1 - MathFuncts.hyperGeometricTest(k, m, n, N);
+        pValueArray.push(pValue);
+      }
+
+      var significantSignatureArray = [];
+
+      var correctedPValues = MathFuncts.multTest.fdr(pValueArray);
+
+      for (i = 0; i < correctedPValues.length; i++) {
+        var correctedPValue = correctedPValues[i].toPrecision(
+          pValueSigDigits);
+
+        if (correctedPValue < self.pValueCutoff) {
+          var signatureId = signatureIds[i];
+          var genes = signatures[signatureId];
+
+          significantSignatureArray.push({
+            'signatureId': signatureId, 'genes': genes,
+            'pValue': correctedPValue
+          });
+        }
+      }
+
+      significantSignatureArray.sort(function(a, b) {
+        return a.pValue - b.pValue;
+      });
+
+      return significantSignatureArray;
+    };
+
+
+    Participation.get(
+      {'mlmodel': self.modelInUrl, 'gene__in': $stateParams.genes, 'limit': 0},
```

Comment:
@rzelayafavila: If `n` on line 104 (see my comment above) is the total number of heavy genes in a certain signature (not limited to the genes in the URL), why we need `gene__in` parameter here? I feel that this parameter should be removed.

Reply:
Ah I see - I think that is true if you want to get ALL participations back from the database and then figure out which participations (and therefore signatures) are relevant to this analysis via JavaScript here.